src/git.rs

```diff
@@ -0,0 +1,300 @@
+use rover_client::query::{graph, subgraph};
+#[derive(Debug)]
+pub struct GitContext {
+    pub branch: Option<String>,
+    pub committer: Option<String>,
+    pub commit: Option<String>,
+    pub message: Option<String>,
+    pub remote_url: Option<String>,
+}
+
+type GraphPushContextInput = graph::push::push_schema_mutation::GitContextInput;
+impl Into<GraphPushContextInput> for GitContext {
+    fn into(self) -> GraphPushContextInput {
+        GraphPushContextInput {
+            branch: self.branch,
+            commit: self.commit,
+            committer: self.committer,
+            remote_url: self.remote_url,
+            message: self.message,
+        }
+    }
+}
+
+type GraphCheckContextInput = graph::check::check_schema_query::GitContextInput;
+impl Into<GraphCheckContextInput> for GitContext {
+    fn into(self) -> GraphCheckContextInput {
+        GraphCheckContextInput {
+            branch: self.branch,
+            commit: self.commit,
+            committer: self.committer,
+            remote_url: self.remote_url,
+            message: self.message,
+        }
+    }
+}
+
+type SubgraphPushContextInput = subgraph::push::push_partial_schema_mutation::GitContextInput;
+impl Into<SubgraphPushContextInput> for GitContext {
+    fn into(self) -> SubgraphPushContextInput {
+        SubgraphPushContextInput {
+            branch: self.branch,
+            commit: self.commit,
+            committer: self.committer,
+            remote_url: self.remote_url,
+            message: self.message,
+        }
+    }
+}
+
+type SubgraphCheckContextInput = subgraph::check::check_partial_schema_query::GitContextInput;
+impl Into<SubgraphCheckContextInput> for GitContext {
+    fn into(self) -> SubgraphCheckContextInput {
+        SubgraphCheckContextInput {
+            branch: self.branch,
+            commit: self.commit,
+            committer: self.committer,
+            remote_url: self.remote_url,
+            message: self.message,
+        }
+    }
+}
+
+impl GitContext {
+    pub fn new(
+        override_branch: Option<String>,
+        override_committer: Option<String>,
+        override_commit: Option<String>,
+        override_remote_url: Option<String>,
+    ) -> Self {
+        let git = git_info::get();
+
+        let branch = if override_branch.is_none() {
```

Comment:
these could all be one liners 

```rust
let branch = override_branch.unwrap_or_else(|| git.current_branch);
```

Reply:
This doesn't quite work, because `git.current_branch` is an `Option<String>`, so the return types clash of it either being a `string` from the unwrapping of `override_branch` or an `Option<String>` from git.

Reply:
you're looking for `or()`! So:
```rust
let branch = override_branch.or(git.current_branch);
```